stdlib/domain.ml

```diff
@@ -187,19 +201,20 @@ let spawn f =
   let term_condition = Condition.create () in
   let term_state = ref Running in
 
-  let at_startup = Atomic.get startup_function in
+  let at_each_spawn = Atomic.get startup_function in
   let body () =
     let result =
       match
         DLS.create_dls ();
         DLS.set_initial_keys pk;
-        at_startup ();
-        f ()
+        at_each_spawn ();
+        let res = f () in
+        do_at_exit ();
+        res
       with
       | x -> Ok x
       | exception ex -> Error ex
     in
-    do_at_exit ();
```

Comment:
If I read the code correctly, now `do_at_exit ()` is only called when the domain callback `f` terminates successfully, but before it would be called even when `f` fails with an exception. Is this correct? Is this change intentional? Is the behavior documented somewhere?

Reply:
Good catch. The intention was captured in the PR message:

> `Domain.at_exit` also used to ignore exceptions. Now, these exceptions are surfaced as the result of domain execution (similar to how `Stdlib.at_exit` works).

I messed up the execution. 

My intention was to match how `Stdlib.at_exit` works. Here's what I shall go for in the fix.

----

If the program terminates normally, and `at_exit` raises an exception, the exception is reported. 

```bash
$ cat exn_exit1.ml 
exception E
exception F

let _ =
  at_exit (fun _ ->
    print_endline "in at_exit callback";
    raise E)
$ ocamlopt -g -o exn_exit1.exe exn_exit1.ml
$ ./exn_exit1.exe 
in at_exit callback
Fatal error: exception Exn_exit1.E
```

----

If the program raises an exception, then the `at_exit` callbacks are run. If the `at_exit` callbacks also raise an exception, then the exception is ignored, and the original exception from the program is reported. 

```bash
$ cat exn_exit2.ml
exception E
exception F

let _ =
  at_exit (fun _ ->
    print_endline "in at_exit callbacks";
    raise E);
  raise F
$ ocamlopt -g -o exn_exit2.exe exn_exit2.ml
$ ./exn_exit2.exe 
in at_exit callback
Fatal error: exception Exn_exit2.F
```

The exceptional behaviour of `Stdlib.at_exit` callbacks is undocumented and is known to have issues (see https://github.com/ocaml/ocaml/issues/11221). Hence, it is better not to aim to find the right semantics for `Domain.at_exit` now. My aim is to follow the semantics of `Stdlib.at_exit`.

Reply:
Done. Please see the changes in the `spawn` function in `domain.ml`.